src/install/dracut-install.c

```diff
@@ -1543,17 +1650,21 @@ static int install_dependent_module(struct kmod_module *mod, int *err)
                 _cleanup_kmod_module_unref_list_ struct kmod_list *modlist = NULL;
                 _cleanup_kmod_module_unref_list_ struct kmod_list *modpre = NULL;
                 _cleanup_kmod_module_unref_list_ struct kmod_list *modpost = NULL;
+                _cleanup_destroy_hashmap_ Hashmap *suppliers = hashmap_new(string_hash_func, string_compare_func);
                 log_debug("dracut_install '%s' '%s' OK", path, &path[kerneldirlen]);
                 install_firmware(mod);
-                modlist = kmod_module_get_dependencies(mod);
-                *err = install_dependent_modules(modlist);
+                *err = find_suppliers_for_module(ctx, suppliers, name);
```

Comment:
if this really is *so much slower* (and that's a fair cop, you're walking a considerable hunk of sysfs for *every* module, AIUI), have you tried walking sysfs *once* and collecting all the module dependencies, then just consulting the resulting map here?

Reply:
Thanks for the suggestion, I just tried it and it seems to speed things up considerably! I'll clean it up and send a new version that walks the sysfs only once.

I also found a bug in my code: if one of the suppliers in a supplier chain is not a module (ie. is built-in), the suppliers it depends on are not processed. So I'll fix that in the next version as well.